### DISCUSSION TITLE
bundle adapters in `hattip` to ease adoption

### DISCUSSION THREAD

**Discussion by samuelstroschein** [hattipjs/hattip#177] [Ideas]
## Problem 

1. Hattip only has a few hundred downloads on NPM making the decision to adopt hattip harder.
2. Adapters are separate packages. Developers have to download multiple `hattip` dependencies slightly lowering DX but, more importantly, spreading the downloads of Hattip across multiple packages. 

## Sugesstion

Bundle all adapters in `hattip` like `hattip/adapter/node`. 

- the npm downloads of all hattip packages are bundled, leading to more downloads and thereby easing adoption
- the deployment pipeline becomes easier because only one package needs to be deployed

We had a similar discussion for inlang https://github.com/inlang/inlang/pull/460#discussion_r1143781116

**Comment by cyco130** [comment]
Hi we're working on a [Vite based CLI](https://github.com/hattipjs/hattip/tree/main/packages/cli/vite) that will bring a very convenient DX. It's only semi-functional right now but when it's ready in a few days, it will feel like using one of Vite-based metaframeworks like SvelteKit, Rakkas, or Astro, minus the frameworK.

There will also be a project initializer (`npm init hattip`) that will allow you to create a project either using the Vite-based CLI or without using any build tool for targets that support it.

I'll close this when we release the CLI.

**Comment by samuelstroschein** [comment]
The CLI won't address the outlined problems :) 

If the CLI is yet another package on NPM, the downloads will be split among yet another npm package. You can bundle everything hattip related in one npm package to incentivise people to adopt hattip due to higher downloads on NPM. 

Thinking out loud here on how to drive adoption for hattip

**Comment by brillout** [comment]
We've plans to significantly increase the number of HatTip users. I think the number of npm downloads won't be an issue then.

**Comment by cyco130** [comment]
> The CLI won't address the outlined problems :)

Yes, but the `hattip` package will. My wires got crossed between this and [the other conversation](https://github.com/hattipjs/hattip/issues/55) :)

**Comment by samuelstroschein** [comment]
> We've plans to significantly increase the number of HatTip users. I think the number of npm downloads won't be an issue then. 

We are interested. Running into multiple issues with express right now. Our architecture is unusual. One server runs everything. Modules export handlers that are imported by the server. So kinda a monolith but not really. Most things are split by module. 

Large downside: 

- how to get HMR for external dependencies in a monorepo?

**Comment by brillout** [comment]
Vite handles linked npm packages as `ssr.noExternal`, so AFAICT HMR should work just fine with your setup.
